code/controllers/subsystem/zaps.dm

```diff
@@ -0,0 +1,382 @@
+				//while still allowing common code to run before hand
+				closest_type = BIKE
+				closest_atom = B
+
+		else if(closest_type >= COIL)
+			continue //no need checking these other things
+
+		else if(istype(A, /obj/machinery/power/energy_accumulator/tesla_coil))
+			if(!HAS_TRAIT(A, TRAIT_BEING_SHOCKED))
+				closest_type = COIL
+				closest_atom = A
+
+		else if(closest_type >= ROD)
+			continue
+
+		else if(istype(A, /obj/machinery/power/energy_accumulator/grounding_rod))
+			closest_type = ROD
+			closest_atom = A
+
+		else if(closest_type >= RIDE)
+			continue
+
+		else if(istype(A,/obj/vehicle/ridden))
+			var/obj/vehicle/ridden/R = A
+			if(R.can_buckle && !HAS_TRAIT(R, TRAIT_BEING_SHOCKED))
+				closest_type = RIDE
+				closest_atom = A
+
+		else if(closest_type >= LIVING)
+			continue
+
+		else if(isliving(A))
+			var/mob/living/L = A
+			if(L.stat != DEAD && !HAS_TRAIT(L, TRAIT_TESLA_SHOCKIMMUNE) && !HAS_TRAIT(L, TRAIT_BEING_SHOCKED))
+				closest_type = LIVING
+				closest_atom = A
+
+		else if(closest_type >= MACHINERY)
+			continue
+
+		else if(ismachinery(A))
+			if(!HAS_TRAIT(A, TRAIT_BEING_SHOCKED))
+				closest_type = MACHINERY
+				closest_atom = A
+
+		else if(closest_type >= BLOB)
+			continue
+
+		else if(istype(A, /obj/structure/blob))
+			if(!HAS_TRAIT(A, TRAIT_BEING_SHOCKED))
+				closest_type = BLOB
+				closest_atom = A
+
+		else if(closest_type >= STRUCTURE)
+			continue
+
+		else if(isstructure(A))
+			if(!HAS_TRAIT(A, TRAIT_BEING_SHOCKED))
+				closest_type = STRUCTURE
+				closest_atom = A
+
+	//Alright, we've done our loop, now lets see if was anything interesting in range
+	if(!closest_atom)
+		return KILL_ZAP
+	//common stuff
+	source.Beam(closest_atom, icon_state="lightning[rand(1,12)]", time = 5)
+	var/zapdir = get_dir(source, closest_atom)
+	if(zapdir)
+		. = zapdir
+
+	var/next_range = 2
+	if(closest_type == COIL)
+		next_range = 5
+
+	if(closest_type == LIVING)
+		var/mob/living/closest_mob = closest_atom
+		ADD_TRAIT(closest_mob, TRAIT_BEING_SHOCKED, WAS_SHOCKED)
+		addtimer(TRAIT_CALLBACK_REMOVE(closest_mob, TRAIT_BEING_SHOCKED, WAS_SHOCKED), 1 SECONDS)
+		var/shock_damage = (zap_flags & ZAP_MOB_DAMAGE) ? (min(round(power/600), 90) + rand(-5, 5)) : 0
+		closest_mob.electrocute_act(shock_damage, source, 1, SHOCK_TESLA | ((zap_flags & ZAP_MOB_STUN) ? NONE : SHOCK_NOSTUN))
+		if(issilicon(closest_mob))
+			var/mob/living/silicon/S = closest_mob
+			if((zap_flags & ZAP_MOB_STUN) && (zap_flags & ZAP_MOB_DAMAGE))
+				S.emp_act(EMP_LIGHT)
+			next_range = 7 // metallic folks bounce it further
+		else
+			next_range = 5
+		power /= 1.5
+
+	else
+		power = closest_atom.zap_act(power, zap_flags)
+
+
+	if(prob(20))//I know I know
+		var/list/shocked_copy = shocked_targets.Copy()
+		tesla_zap(closest_atom, next_range, power * 0.5, zap_flags, shocked_copy, TRUE)//Normally I'd copy here so grounding rods work properly, but it fucks with movement
```

Comment:
This might change behavior compared to the original when it was a recursive proc. I'm not really sure how to get around this. @LemonInTheDark